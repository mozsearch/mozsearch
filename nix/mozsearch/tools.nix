{
  rustPlatform,
  runCommandLocal,
  pkgconf,
  cmake,
  openssl,
}: let
  subdir = "tools";
in
  rustPlatform.buildRustPackage {
    pname = "mozsearch-tools";
    version = "unstable";

    src = runCommandLocal "mozsearch-tools-source" {} ''
      mkdir -p $out
      cp -r ${../../tools} $out/tools
      cp -r ${../../deps} $out/deps
    '';

    buildAndTestSubdir = subdir;
    cargoRoot = subdir;

    cargoLock = {
      lockFile = ../../${subdir}/Cargo.lock;
      outputHashes = {
        "traitobject-0.1.0" = "sha256-hnCMcSLblsa0BTRxV/Q7wE1C2KByIf51jYvZGoKFV4M=";
      };
    };

    nativeBuildInputs = [
      pkgconf
      cmake
    ];

    buildInputs = [
      openssl
    ];
  }
