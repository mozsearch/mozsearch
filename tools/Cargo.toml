[package]
name = "tools"
version = "0.1.0"
authors = ["Bill McCloskey <billm@mozilla.com>"]

[dependencies]
chrono = "0.2"
rustc-serialize = "0.3.18"
git2 = "0.7"
lazy_static = "0.2.1"
hyper = "0.10"
log = "0.4.0"
env_logger = "0.5.13"
regex = "1"
ipdl_parser = { path = "./ipdl_parser" }
getopts = "0.2.14"
memmap = "0.5.0"
linkify = "0.2.0"
rls-analysis = "0.17"
rls-data = "0.19"
clap = "2"

# Build release mode with line number info for easier debugging when
# we hit panics in production
[profile.release]
debug = 1
