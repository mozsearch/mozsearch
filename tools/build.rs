fn main() -> Result<(), Box<dyn std::error::Error>> {
    // right now we only use the livegrep proto and not the config proto
    tonic_build::configure().build_server(false).compile(
        &["../deps/livegrep/src/proto/livegrep.proto"],
        &["../deps/livegrep/"],
    )?;
    Ok(())
}
