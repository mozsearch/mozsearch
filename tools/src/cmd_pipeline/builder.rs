use crate::{
    cmd_pipeline::{
        cmd_prod_filter::ProductionFilterCommand, cmd_query::QueryCommand,
        cmd_search_text::SearchTextCommand, PipelineCommand,
    },
    structopt::StructOpt,
};
use tracing::{trace, trace_span};
use url::Url;

use crate::{
    abstract_server::{
        make_local_server, make_remote_server, ErrorDetails, ErrorLayer, Result, ServerError,
    },
    cmd_pipeline::parser::{Command, OutputFormat, ToolOpts},
};

use super::{cmd_search::SearchCommand, cmd_search_files::SearchFilesCommand, cmd_crossref_expand::CrossrefExpandCommand};
use super::cmd_show_html::ShowHtmlCommand;
use super::cmd_traverse::TraverseCommand;
use super::{
    cmd_crossref_lookup::CrossrefLookupCommand, cmd_filter_analysis::FilterAnalysisCommand,
    cmd_graph::GraphCommand, cmd_merge_analyses::MergeAnalysesCommand,
    cmd_search_identifiers::SearchIdentifiersCommand,
};

use super::interface::ServerPipeline;

pub fn fab_command_from_opts(opts: ToolOpts) -> Result<Box<dyn PipelineCommand>> {
    match opts.cmd {
        Command::CrossrefExpand(ce) => Ok(Box::new(CrossrefExpandCommand { args: ce })),

        Command::CrossrefLookup(cl) => Ok(Box::new(CrossrefLookupCommand { args: cl })),

        Command::FilterAnalysis(fa) => Ok(Box::new(FilterAnalysisCommand { args: fa })),

        Command::Graph(g) => Ok(Box::new(GraphCommand { args: g })),

        Command::MergeAnalyses(ma) => Ok(Box::new(MergeAnalysesCommand { args: ma })),

        Command::ProductionFilter(pf) => Ok(Box::new(ProductionFilterCommand { args: pf })),

        Command::Query(q) => Ok(Box::new(QueryCommand { args: q })),

        Command::Search(q) => Ok(Box::new(SearchCommand { args: q })),

        Command::SearchFiles(sf) => Ok(Box::new(SearchFilesCommand { args: sf })),

        Command::SearchIdentifiers(si) => Ok(Box::new(SearchIdentifiersCommand { args: si })),

        Command::SearchText(st) => Ok(Box::new(SearchTextCommand { args: st })),

        Command::ShowHtml(sh) => Ok(Box::new(ShowHtmlCommand { args: sh })),

        Command::Traverse(t) => Ok(Box::new(TraverseCommand { args: t })),
    }
}

/// Build a command pipeline from a shell-y string where we use pipe boundaries
/// to delineate the separate pipeline steps.
///
/// The shell-words module is used to parse `arg_str` into shell words, which we
/// then break into separate sub-commands whenever we see a `|`.  We then pass
/// these sub-commands to the structopt parsing `from_iter` method, taking care
/// to stuff our binary name into the first arg.
pub fn build_pipeline(bin_name: &str, arg_str: &str) -> Result<(ServerPipeline, OutputFormat)> {
    let span = trace_span!("build_pipeline", arg_str);
    let _span_guard = span.enter();

    let all_args = match shell_words::split(arg_str) {
        Ok(parsed) => parsed,
        Err(err) => {
            return Err(ServerError::StickyProblem(ErrorDetails {
                layer: ErrorLayer::BadInput,
                message: err.to_string(),
            }));
        }
    };

    let mut server_kind = "none";
    let mut server = None;
    let mut output_format = None;
    let mut first_time = true;

    let mut commands: Vec<Box<dyn PipelineCommand>> = vec![];

    for arg_slices in all_args.split(|v| v == "|") {
        let mut fake_args = vec![bin_name.to_string()];
        fake_args.extend(arg_slices.iter().cloned());

        let opts = match ToolOpts::from_iter_safe(fake_args) {
            Ok(opts) => opts,
            Err(err) => {
                return Err(ServerError::StickyProblem(ErrorDetails {
                    layer: ErrorLayer::BadInput,
                    message: err.to_string(),
                }));
            }
        };
        //println!("Pipeline segment: {:?}", opts);

        if first_time {
            (server_kind, server) = match Url::parse(&opts.server) {
                Ok(url) => ("remote", Some(make_remote_server(url, &opts.tree)?)),
                Err(_) => ("local", Some(make_local_server(&opts.server, &opts.tree)?)),
            };
            output_format = Some(opts.output_format.clone());
            first_time = false;
        }

        trace!(cmd = ?opts.cmd);
        commands.push(fab_command_from_opts(opts)?);
    }

    Ok((
        ServerPipeline {
            server_kind: server_kind.to_string(),
            server: server.unwrap(),
            commands,
        },
        output_format.unwrap(),
    ))
}

/*
pub fn build_pipeline_graph(query: QueryPipelineGroupBuilder) -> ServerPipelineGraph {

}
*/
