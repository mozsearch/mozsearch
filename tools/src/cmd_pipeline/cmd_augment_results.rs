use std::{cell::Cell, collections::HashMap, rc::Rc};

use async_trait::async_trait;
use clap::Parser;
use lol_html::{element, HtmlRewriter, Settings};
use ustr::UstrMap;

use super::interface::{PipelineCommand, PipelineValues};
use crate::abstract_server::{
    AbstractServer, ErrorDetails, ErrorLayer, HtmlFileRoot, Result, ServerError,
};

/// Augment a FlattenedResultsBundle by scraping the rendered HTML output files
/// for lines of interest plus any context, plus applying any predicates that
/// run against data baked into the output file (like coverage data or history
/// data).
///
/// It's possible the performance for this will be horrible, although it may be
/// possible to throw some of the following at this model:
/// - Having the HTML line extraction run as multiple parallel tasks.
/// - Increasing VM RAM to allow for more caching.
/// - Switching to a VM with local SSD for better I/O.
/// - Pre-computing line offsets (in the uncompresed HTML, noting that we do
///   gzip the HTML and probably still want to seek through that) so that we can
///   reduce the amount of HTML parsing required (even if we still have to seek
///   through a lot).
/// - Create some alternate on-disk representation for the rendered HTML files
///   that's intentionally distinct from the pre-gzipped files.  This could be
///   a different compression format and/or sharded files (maybe with duplicate
///   data at the start/end so we never have to span shards normally).
///
/// Alternately, we might:
/// - Run a post-file-rendering phase and effectively re-compute the crossref
///   database with HTML baked in and some number of extra lines of context?
/// - Have the crossref database include some extra context and have tokenizer
///   state included at the first line point so the tokenizer can do a bare
///   bones syntax highlighting.
#[derive(Debug, Parser)]
pub struct AugmentResults {
    /// Lines of context before a hit.
    #[clap(short, long, value_parser, default_value = "0")]
    before: u32,

    /// Lines of context after a hit.
    #[clap(short, long, value_parser, default_value = "0")]
    after: u32,
}

#[derive(Debug)]
pub struct AugmentResultsCommand {
    pub args: AugmentResults,
}

#[async_trait]
impl PipelineCommand for AugmentResultsCommand {
    async fn execute(
        &self,
        server: &Box<dyn AbstractServer + Send + Sync>,
        input: PipelineValues,
    ) -> Result<PipelineValues> {
        let mut results = match input {
            PipelineValues::FlattenedResultsBundle(frb) => frb,
            _ => {
                return Err(ServerError::StickyProblem(ErrorDetails {
                    layer: ErrorLayer::ConfigLayer,
                    message: "augment-resultst needs a FlattenedResultsBundle".to_string(),
                }));
            }
        };

        // ### Build up map of HTML lines
        //
        // This bit could potentially run in parallel.  We're not going to yet
        // because:
        // - Wanna see how bad it is.
        // - Don't wanna have this experimental thing steal all the resources
        //   from the non-experimental router.py and web-server.rs yet!

        let mut path_line_contents: UstrMap<HashMap<u32, String>> = UstrMap::default();

        for (path, lines_to_show) in
            results.compute_path_line_sets(self.args.before, self.args.after)
        {
            // XXX Doing this as a single string received in a lump is fine for
            // our testing use-case, but this may need to be reconsidered in
            // production.  Or maybe production really wants the performance?
            // Production certainly should have the RAM for our known worst
            // case scenarios.
            let html_str = server
                .fetch_html(HtmlFileRoot::FormattedFile, &path)
                .await?;

            let file_lines = path_line_contents.entry(path).or_default();

            // ### HTML Extraction: What We Want
            //
            // We want the full line container which looks like:
            // - div id="line-N" class="source-line-with-number" role="row"
            //   - div role="cell"
            //     - div class="cov-strip cov-uncovered cov-known"
            //   - div role="cell"
            //     - div class="blame-strip c2" data-blame="..."
            //   - div role="cell" class="line-number" data-line-number="N"
            //   - code role="cell" class="source-line"
            //     - ex: span class="sync_comment"
            //     - ex: span class="syn_def syn_type" data-symbols="..." data-i
            //
            // ### HTML Extraction Low Level Details
            //
            // Until https://github.com/cloudflare/lol-html/issues/40 or
            // the spin-off https://github.com/cloudflare/lol-html/issues/78
            // are implemented, lol_html doesn't explicitly provide a way to
            // derive the value of an element.
            //
            // So we attempt a hack where we use a custom output sink that is
            // kept aware of where we are in the file.  The good news is that
            // since lol_html is oriented around minimal memory allocation, we
            // can generally control when flushes happen.

            let mut writing_line: u32 = 0;
            let cur_line = Cell::new(0u32);
            let want_cur_line = Cell::new(false);
            let suppressing = Rc::new(Cell::new(false));
            let nesting_suppress = suppressing.clone();

            let mut buf = vec![];

            let mut rewrite = HtmlRewriter::new(
                Settings {
                    element_content_handlers: vec![
                        element!(r#"div.nesting-container"#, move |el| {
                            nesting_suppress.set(true);
                            let end_suppress = nesting_suppress.clone();
                            el.on_end_tag(move |_end| {
                                end_suppress.set(true);
                                Ok(())
                            })?;
                            Ok(())
                        }),
                        element!(r#"div.source-line-with-number"#, |el| {
                            suppressing.set(false);
                            if let Some(id_str) = el.get_attribute("id") {
                                let id_parts: Vec<&str> = id_str.split("-").collect();
                                if id_parts.len() == 2 && id_parts[0] == "line" {
                                    let lno = id_parts[1].parse().unwrap_or(0);
                                    cur_line.set(lno);
                                    want_cur_line.set(lines_to_show.contains(&lno));
                                }
                            }

                            Ok(())
                        }),
                    ],
                    ..Settings::default()
                },
                |c: &[u8]| {
                    if suppressing.get() {
                        return;
                    }

                    // We were actively writing and potentially have some
                    // buffer.
                    if writing_line > 0 {
                        // We're done writing; flush!
                        if cur_line.get() != writing_line {
                            file_lines
                                .insert(writing_line, String::from_utf8_lossy(&buf).to_string());
                            writing_line = 0;
                            buf.clear();
                        }
                        // We're still writing!
                        else {
                            // Write into the buffer and then leave, because we
                            // don't need to consider switching into writing, as
                            // we're still here.
                            buf.extend_from_slice(c);
                            return;
                        }
                    }
                    // We either closed out writing or weren't writing.  But now
                    // we need to see if we should be writing!
                    if cur_line.get() > 0 && want_cur_line.get() {
                        writing_line = cur_line.get();
                        buf.extend_from_slice(c);
                    }
                    // Otherwise, this wasn't interesting.
                },
            );

            rewrite.write(html_str.as_bytes()).unwrap();
            rewrite.end().unwrap();
        }

        // ## Ingest the new lines.
        results.ingest_html_lines(&path_line_contents, self.args.before, self.args.after);

        Ok(PipelineValues::FlattenedResultsBundle(results))
    }
}
