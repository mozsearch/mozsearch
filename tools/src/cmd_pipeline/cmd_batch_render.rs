use async_trait::async_trait;
use serde_json::{to_value};
use clap::Args;

use super::{interface::{JsonValue, PipelineCommand, PipelineValues}, builder::build_pipeline_graph};
use crate::{
    abstract_server::{AbstractServer, Result}, query::chew_query::chew_query,
};

#[derive(Debug, Args)]
pub struct BatchRender {
    /// Preconfigured rendering task.  This could be an enum or sub-command, but
    /// for now we're just going for strings.
    task: String,
}

#[derive(Debug)]
pub struct BatchRenderCommand {
    pub args: BatchRender,
}

#[async_trait]
impl PipelineCommand for BatchRenderCommand {
    async fn execute(
        &self,
        server: &Box<dyn AbstractServer + Send + Sync>,
        _input: PipelineValues,
    ) -> Result<PipelineValues> {
        let pipeline_plan = chew_query(&self.args.query)?;

        if self.args.dump_pipeline {
            return Ok(PipelineValues::JsonValue(JsonValue { value: to_value(pipeline_plan)? }));
        }

        let graph = build_pipeline_graph(server.clonify(), pipeline_plan)?;

        let result = graph.run(true).await;
        result
    }
}
