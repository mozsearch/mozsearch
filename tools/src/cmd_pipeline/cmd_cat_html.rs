use std::{cell::Cell, rc::Rc};
use async_trait::async_trait;
use lol_html::{element, rewrite_str, HtmlRewriter, RewriteStrSettings, Settings, html_content::ContentType};
use clap::Args;

use super::interface::{PipelineCommand, PipelineValues, TextFile};
use crate::{
    abstract_server::{AbstractServer, HtmlFileRoot, Result},
};

/// Dump the contents of a HTML file for a (source) file or rendered directory
/// listing from disk in its entirety, applying minimal normalization to
/// compensate for datestamps or specific revision data.
///
/// Intended exclusively for regression testing and in particular to provide
/// coverage for directory listings and semi-generated files like "help.html"
/// and "settings.html" where we want to audit changes to the file in their
/// entirety but we don't want to have N copies of the searchfox HTML super
/// structure.  We would expect to use this command for at most a few source
/// files to validate where the source listing joins to the HTML super structure
/// but would expect to use "show-html" for most "checks" so that they can be
/// much more targeted (than having potentially massive diffs that are
/// constantly including the entirety of a generated page with many irrelevant
/// changes to the specific purpose of the check).
///
/// Differs from show-html which is about excerpting source lines and which has
/// a separate "prod-filter" helper for production "checks".
#[derive(Debug, Args)]
pub struct CatHtml {
    /// Tree-relative source file path or directory.
    #[clap(value_parser)]
    file: String,

    /// Is this a directory's HTML we want (instead of a source file)?
    #[clap(short, long, action)]
    dir: bool,

    /// Is this a template's HTML we want?
    #[clap(short, long, action)]
    template: bool,

    /// Use a CSS selector to limit the returned portion of the document.  This
    /// can be useful to focus a test and make diffs easier to understand.
    #[clap(short, long, value_parser)]
    select: Option<String>,
}

#[derive(Debug)]
pub struct CatHtmlCommand {
    pub args: CatHtml,
}

// HTML normalization of our expected entire HTML files:
// - "This page was generated by Searchfox DATETIME": We wrap the
//   datetime in a spam with class "pretty-date" and attribute with key
//   "data-datetime".  We currently normalize by replacing it with a span
//   `<span>NORMALIZED</span>` which loses the extra attributes but we don't
//   care about that level of fidelity.
fn norm_html_file(s: String) -> String {
    let element_content_handlers = vec![
        element!(
            r#"span.pretty-date"#,
            |el| {
                el.replace("<span>NORMALIZED</span>", ContentType::Html);
                Ok(())
            }
        ),
    ];

    rewrite_str(
        &s,
        RewriteStrSettings {
            element_content_handlers,
            ..RewriteStrSettings::default()
        },
    )
    .unwrap()
}

fn extract_html_snippet(html_str: String, selector: &str) -> String {
    let mut excerpts = vec![];

    let suppressing = Rc::new(Cell::new(true));
    let sink_suppressing = suppressing.clone();

    let mut buf = vec![];

    let synthetic_closing = Rc::new(Cell::new(None));
    let sink_closing = synthetic_closing.clone();

    let mut rewrite = HtmlRewriter::new(
        Settings {
            element_content_handlers: vec![
                element!(
                    selector,
                    move |el| {
                        suppressing.set(false);
                        let end_suppress = suppressing.clone();
                        let end_closing = synthetic_closing.clone();
                        el.on_end_tag(move |end| {
                            end_closing.set(Some(format!("</{}>", end.name())));
                            end_suppress.set(true);
                            Ok(())
                        })?;
                        Ok(())
                    }
                ),
            ],
            ..Settings::default()
        },
        |c: &[u8]| {
            if sink_suppressing.get() {
                if let Some(closing) = sink_closing.take() {
                    buf.extend_from_slice(closing.as_bytes());
                }
                // Flush if this was apparently a transition from accumulating
                // into our buffer.
                if buf.len() > 0 {
                    excerpts.push(String::from_utf8_lossy(&buf).to_string());
                    buf.clear();
                }
                return;
            }

            buf.extend_from_slice(c);
        },
    );

    rewrite.write(html_str.as_bytes()).unwrap();
    rewrite.end().unwrap();

    excerpts.join("\n")
}

#[async_trait]
impl PipelineCommand for CatHtmlCommand {
    async fn execute(
        &self,
        server: &Box<dyn AbstractServer + Send + Sync>,
        _input: PipelineValues,
    ) -> Result<PipelineValues> {
        let root = if self.args.dir {
            HtmlFileRoot::FormattedDir
        } else if self.args.template {
            HtmlFileRoot::FormattedTemplate
        } else {
            HtmlFileRoot::FormattedFile
        };
        let mut html_str = server.fetch_html(root, &self.args.file).await?;

        if let Some(selector) = &self.args.select {
            html_str = extract_html_snippet(html_str, selector);
        }

        Ok(PipelineValues::TextFile(TextFile {
            mime_type: "text/html".to_string(),
            contents: norm_html_file(html_str),
        }))

    }
}
