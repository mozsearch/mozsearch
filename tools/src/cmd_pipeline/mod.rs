extern crate clap;
extern crate structopt;

pub mod builder;
pub mod interface;
pub mod parser;

mod cmd_filter_analysis;
mod cmd_show_html;

pub use builder::{build_pipeline};
pub use interface::{PipelineCommand, PipelineValues};
