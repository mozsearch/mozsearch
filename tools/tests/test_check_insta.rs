use regex::Regex;
use tokio::fs::{read_dir, read_to_string};
use tools::cmd_pipeline::{build_pipeline, PipelineValues};

/// Glob-style insta test where we process all of the searchfox-tool command
/// lines in TREE/checks/inputs and output the results of those pipelines to
/// TREE/checks/snapshots using `insta` which provides diff functionality.
///
/// This very dubiously currently relies on having an environment variable
/// CHECK_ROOT defined to tell us where the TREE is.  One might wonder whether
/// this should actually be a test at all or whether it should be its own binary
/// or maybe searchfox-tool should know how to do this or what.
///
/// The reality is that we expect this to be invoked indirectly via
/// `check-index.sh` and never directly triggered via `cargo test`, so... yeah,
/// maybe we could do this in a more clean fashion.  Better opinions accepted!
///
/// `insta` does provide support for binding settings using an `async` function,
/// but its "glob" mechanism does not support `async` so we attempt to reproduce
/// the subset we need from tokio::fs.  (We don't need to use tokio for this,
/// but since we've already started down that road, we stay on the road.  The
/// use of tokio for this is separate from the async limitations on insta's glob
/// which necessitate us doing our own file-finding.)
#[tokio::test(flavor = "multi_thread")]
async fn test_check_glob() -> Result<(), std::io::Error> {
    if let Ok(check_root) = std::env::var("CHECK_ROOT") {
        let input_path = format!("{}/inputs", check_root);
        let snapshot_path = format!("{}/snapshots", check_root);

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path(snapshot_path);
        settings.set_prepend_module_to_snapshot(false);

        // ## Figure out the list of input files and sort them.
        let mut dir = read_dir(input_path).await?;
        let mut input_names = Vec::new();

        while let Some(child) = dir.next_entry().await? {
            if child.metadata().await?.is_file() {
                input_names.push(
                    child
                        .path()
                        .file_name()
                        .unwrap()
                        .to_str()
                        .unwrap()
                        .to_string(),
                );
            }
        }

        // Regexp to help with stripping coverage and blame lines which are not
        // expected to be stable in production for HTML lines.
        //
        // TODO: Make this something that doesn't impact the mozsearch test
        // repository, since its coverage data is expected to always be canned.
        let strip_stripper =
            Regex::new(r#"(?m)^  <div role="cell"><div class="(?:cov|blame)-strip.+</div></div>\n"#)
                .unwrap();

        for input_name in input_names {
            let input_path = format!("{}/inputs/{}", check_root, input_name);
            settings.set_input_file(&input_path);
            settings.set_snapshot_suffix(input_name);

            settings
                .bind_async(async {
                    let command = read_to_string(input_path).await.unwrap();

                    let pipeline = match build_pipeline(&"searchfox-tool", &command) {
                        Ok((pipeline, _)) => pipeline,
                        Err(err) => {
                            insta::assert_snapshot!(format!("Pipeline Build Error: {:?}", err));
                            return;
                        }
                    };
                    let results = pipeline.run().await;

                    match results {
                        Ok(PipelineValues::Void) => {
                            insta::assert_snapshot!("void");
                        }
                        Ok(PipelineValues::HtmlExcerpts(he)) => {
                            let mut aggr_str = String::new();
                            for file_excerpts in he.by_file {
                                //println!("HTML excerpts from: {}", file_excerpts.file);
                                for str in file_excerpts.excerpts {
                                    // Normalize out the blame/coverage lines.
                                    // See the regexp def for more info.
                                    aggr_str += &strip_stripper.replace_all(&str, "");
                                }
                            }
                            insta::assert_snapshot!(&aggr_str);
                        }
                        Ok(PipelineValues::JsonRecords(jr)) => {
                            let mut json_results = vec![];
                            for file_records in jr.by_file {
                                json_results.extend(file_records.records);
                            }
                            insta::assert_json_snapshot!(&json_results);
                        }
                        Err(err) => {
                            insta::assert_snapshot!(format!("Pipeline Error: {:?}", err));
                        }
                    }
                })
                .await;
        }
    }

    Ok(())
}
