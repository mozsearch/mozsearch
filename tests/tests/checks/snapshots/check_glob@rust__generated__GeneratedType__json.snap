---
source: tests/test_check_insta.rs
expression: "&json_results"

---
[
  {
    "kind": "def",
    "loc": "00002:11-24",
    "pretty": "simple::build_time_generated::GeneratedType",
    "sym": "simple::build_time_generated::GeneratedType",
    "target": 1
  }
]
